src/browser/modules/Editor/ProjectFilesButton.tsx

```diff
@@ -0,0 +1,218 @@
+  PROJECT_FILES_MOUNTED,
+  PROJECT_FILES_UNMOUNTED
+} from 'browser/modules/Sidebar/project-files.constants'
+import { useMutation } from '@apollo/client'
+import { isWindows } from '../App/keyboardShortcuts'
+
+const PROJECT_EDITOR_BUTTON_TITLE = 'Project File'
+const PROJECT_SAVE_EDITOR_BUTTON_TITLE = 'Save Project File'
+const PROJECT_EDIT_EDITOR_BUTTON_TITLE = 'Edit Project File'
+const PROJECT_FILES_DRAWER_NAME = 'project files'
+
+interface ProjectFilesButtonProps {
+  width?: number
+  editorValue: () => string
+  bus: Bus
+  toggleDrawer: () => void
+  isDrawerOpen: boolean
+  isRelateAvailable: boolean
+  projectId: string
+}
+
+type ActiveRelateFile = Omit<ProjectFile, 'downloadToken'>
+
+const ProjectFileButton = ({
+  width = 24,
+  editorValue,
+  bus,
+  toggleDrawer,
+  isDrawerOpen,
+  isRelateAvailable,
+  projectId
+}: ProjectFilesButtonProps): JSX.Element | null => {
+  const [addFile, { error }] = useMutation(ADD_PROJECT_FILE)
+  const [isEditMode, setEditMode] = useState(false)
+  const [isSaveModeClick, setSaveModeClick] = useState(false)
+  const [isProjectFilesMounted, setProjectFilesMounted] = useState(false)
+  const [
+    activeRelateFile,
+    setActiveRelateFile
+  ] = useState<ActiveRelateFile | null>(null)
+
+  useEffect(() => {
+    let isStillMounted = true
+
+    // when a saved Project Script or Local Cache Script is clicked
+    // not sure at this point which it could be
+    bus &&
+      bus.take(EDIT_CONTENT, () => {
+        if (isStillMounted) {
+          setEditMode(false)
+          setActiveRelateFile(null)
+        }
+      })
+    // only when a Project Script is clicked
+    bus &&
+      bus.take(SELECT_PROJECT_FILE, projectCypherFile => {
+        if (isStillMounted) {
+          setEditMode(true)
+          setActiveRelateFile(projectCypherFile)
+        }
+      })
+    // when a non-Project File action sets content in the editor
+    bus &&
+      bus.take(SET_CONTENT, () => {
+        if (isStillMounted) {
+          setEditMode(false)
+          setActiveRelateFile(null)
+        }
+      })
+    // Project Files is mounted
+    bus &&
+      bus.take(PROJECT_FILES_MOUNTED, () => {
+        if (isStillMounted) {
+          setProjectFilesMounted(true)
+        }
+      })
+    // Prject Files is unmounted
+    bus &&
+      bus.take(PROJECT_FILES_UNMOUNTED, () => {
+        if (isStillMounted) {
+          setProjectFilesMounted(false)
+        }
+      })
+
+    return () => {
+      isStillMounted = false
+    }
+  }, [])
+
+  useEffect(() => {
+    // to report an error in the EditorFrame
+    bus && bus.send(PROJECT_FILE_ERROR, '')
+  }, [error])
+
+  useEffect(() => {
+    // bus sends message too quickly before ProjectFiles is mounted,
+    // so need to ensure ProjectsFile is mounted first.
+    // isDrawerOpen is not reliable in this case.
+    if (isSaveModeClick && isProjectFilesMounted) {
+      bus.send(SAVE_PROJECT_FILE, editorValue())
+      setSaveModeClick(false)
+    }
```

Comment:
second parameter? or switch to useCallback?

Reply:
ah right, I think I know how I could use the `useCallback` now. Ill give it a go.

Reply:
tried `useCallback` but still a race condition so added a dependency array instead